pkg/controller/federatedcluster/controller.go

```diff
@@ -160,82 +150,146 @@ func (cc *ClusterController) Run(stopChan <-chan struct{}) {
+		cc.mu.RUnlock()
+		if clusterStore == nil {
+			glog.Warningf("Failed to retrieve stored data for cluster %s", cluster.Name)
 			continue
 		}
-		clusterClient, clientFound := cc.clusterKubeClientMap[cluster.Name]
-		clusterStatusOld, statusFound := cc.clusterStatusMap[cluster.Name]
-		cc.mu.RUnlock()
 
-		if !clientFound {
-			glog.Warningf("Failed to get client for cluster %s", cluster.Name)
-			continue
+		wg.Add(1)
+		go cc.updateIndividualClusterStatus(cluster, clusterStore, &wg)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (cc *ClusterController) updateIndividualClusterStatus(cluster *fedv1a1.FederatedCluster,
+	store *ClusterStore, wg *sync.WaitGroup) {
+	clusterClient := store.clusterKubeClient
+
+	currentClusterStatus := thresholdAdjustedClusterStatus(clusterClient.GetClusterHealthStatus(), store,
+		cc.clusterHealthCheckConfig)
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossClusterServiceDiscovery) {
+		currentClusterStatus = updateClusterZoneAndRegion(currentClusterStatus, store, clusterClient, cluster.Name)
+	}
+
+	store.clusterStatus = currentClusterStatus
+	cluster.Status = *currentClusterStatus
+	if err := cc.client.UpdateStatus(context.TODO(), cluster); err != nil {
+		glog.Warningf("Failed to update the status of cluster: %v, error is : %v", cluster.Name, err)
+	}
+	wg.Done()
+}
+
+func thresholdAdjustedClusterStatus(clusterStatus *fedv1a1.FederatedClusterStatus, store *ClusterStore,
+	clusterHealthCheckConfig util.ClusterHealthCheckConfig) *fedv1a1.FederatedClusterStatus {
+
+	if store.clusterStatus == nil {
+		return clusterStatus
+	}
+
+	threshold := clusterHealthCheckConfig.FailureThreshold
+	if isClusterReady(clusterStatus) {
+		threshold = clusterHealthCheckConfig.SuccessThreshold
+	}
+
+	if store.resultRun < threshold {
+		// Success/Failure is below threshold - leave the probe state unchanged.
+		probeTime := clusterStatus.Conditions[0].LastProbeTime
+		clusterStatus = store.clusterStatus
+		setLastProbeTime(clusterStatus, probeTime)
+	} else {
+		if clusterStatusEqual(clusterStatus, store.clusterStatus) {
+			// preserve the last transition time
+			setLastTransitionTime(clusterStatus, store.clusterStatus.Conditions[0].LastTransitionTime)
 		}
-		clusterStatusNew := clusterClient.GetClusterHealthStatus()
-		if !statusFound {
-			glog.Infof("There is no status stored for cluster: %v before", cluster.Name)
-		} else {
-			hasTransition := false
-			if len(clusterStatusNew.Conditions) != len(clusterStatusOld.Conditions) {
-				hasTransition = true
-			} else {
-				for i := 0; i < len(clusterStatusNew.Conditions); i++ {
-					if !(strings.EqualFold(string(clusterStatusNew.Conditions[i].Type), string(clusterStatusOld.Conditions[i].Type)) &&
-						strings.EqualFold(string(clusterStatusNew.Conditions[i].Status), string(clusterStatusOld.Conditions[i].Status))) {
-						hasTransition = true
-						break
-					}
-				}
-			}
+	}
+
+	if clusterStatusEqual(clusterStatus, store.clusterStatus) {
```

Comment:
how about "isClusterReady(clusterStatus) != isClusterReady(store.clusterStatus)"?
keeping resultRun small while flipping between two different non-ready conditions doesn't make much sense.

Reply:
your comment makes sense. I initially was pondering upon the same for quite some time. Updated to just store Ready or NotReady states. ptal once again.